modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TransactionExpirationRegistry.java

```diff
@@ -0,0 +1,124 @@
+
+class TransactionExpirationRegistry {
+    private static final IgniteLogger LOG = Loggers.forClass(TransactionExpirationRegistry.class);
+
+    private final NavigableMap<HybridTimestamp, Set<InternalTransaction>> txsByExpirationTime = new ConcurrentSkipListMap<>();
+    private final Map<InternalTransaction, HybridTimestamp> expirationTimeByTx = new ConcurrentHashMap<>();
+
+    private final ReadWriteLock watermarkLock = new ReentrantReadWriteLock();
+    private volatile HybridTimestamp watermark = HybridTimestamp.MIN_VALUE;
+
+    void register(InternalTransaction tx, HybridTimestamp txExpirationTime) {
+        if (isExpired(txExpirationTime)) {
+            abortTransaction(tx);
+            return;
+        }
+
+        watermarkLock.readLock().lock();
+
+        try {
+            if (isExpired(txExpirationTime)) {
+                abortTransaction(tx);
+                return;
+            }
+
+            Set<InternalTransaction> txsExpiringAtTs = txsByExpirationTime.computeIfAbsent(
+                    txExpirationTime,
+                    k -> ConcurrentHashMap.newKeySet()
+            );
+            txsExpiringAtTs.add(tx);
+
+            expirationTimeByTx.put(tx, txExpirationTime);
+        } finally {
+            watermarkLock.readLock().unlock();
+        }
+    }
+
+    private boolean isExpired(HybridTimestamp expirationTime) {
+        return expirationTime.compareTo(watermark) <= 0;
+    }
+
+    private static void abortTransaction(InternalTransaction tx) {
+        tx.rollbackAsync().whenComplete((res, ex) -> {
+            if (ex != null) {
+                LOG.error("Transaction abort due to timeout failed [txId={}]", ex, tx.id());
+            }
+        });
+    }
+
+    void expireUpTo(HybridTimestamp expirationTime) {
+        List<Set<InternalTransaction>> transactionSetsToExpire;
+
+        watermarkLock.writeLock().lock();
+
+        try {
+            NavigableMap<HybridTimestamp, Set<InternalTransaction>> headMap = txsByExpirationTime.headMap(expirationTime, true);
+            transactionSetsToExpire = new ArrayList<>(headMap.values());
+            headMap.clear();
+
+            watermark = expirationTime;
+        } finally {
+            watermarkLock.writeLock().unlock();
+        }
+
+        for (Set<InternalTransaction> set : transactionSetsToExpire) {
+            for (InternalTransaction tx : set) {
+                expirationTimeByTx.remove(tx);
+
+                abortTransaction(tx);
+            }
+        }
+    }
+
+    void abortAllRegistered() {
+        expireUpTo(HybridTimestamp.MAX_VALUE);
+    }
+
+    void unregister(InternalTransaction tx) {
+        HybridTimestamp expirationTime = expirationTimeByTx.remove(tx);
+
+        if (expirationTime != null) {
+            txsByExpirationTime.compute(expirationTime, (k, set) -> {
+                if (set == null) {
+                    return null;
+                }
+
+                set.remove(tx);
```

Comment:
Do I understand correctly that most of the time this set will only contain one element? Should we optimize for it?